utils/doclint/cli.js

```diff
@@ -122,26 +122,105 @@ async function run() {
+            return matches.includes(`-${lang}`);
+          })
+          // Standardise naming and remove the filter in the file name
+          .map(filePath => filePath.replace(/(-(js|python|csharp|java))+/, ''))
+          // Internally we merge test-api and test-reporter-api into api.
+          .map(filePath => filePath.replace(/\/(test-api|test-reporter-api)\//, '/api/'))
+          // Strip off the root
+          .map(filePath => filePath.substring(documentationRoot.length + 1))
+
+        /**
+         * @param {string} filePath
+         * @returns {boolean}
+         */
+        function hasDocFile(filePath) {
+          const kIgnoreDocFiles = ['test-assertions.md'];
+          // We generate it inside the generator
+          if (kIgnoreDocFiles.includes(filePath))
+            return true;
+          return relevantMarkdownFiles.some(other => {
+            if (other === filePath)
+              return true;
+            if (other === path.join('api', filePath))
+              return true;
+            if (lang === 'js') {
+              if (other === path.join('test-api', filePath))
+                return true;
+              if (other === path.join('test-reporter-api', filePath))
+                return true;
+            }
+            return false;
+          });
+        }
+
+        for (const filePath of getAllMarkdownFiles(documentationRoot)) {
           if (langs.some(other => other !== lang && filePath.endsWith(`-${other}.md`)))
             continue;
           const data = fs.readFileSync(filePath, 'utf-8');
-          documentation.renderLinksInText(md.filterNodesForLanguage(md.parse(data), lang));
+          const rootNode = md.filterNodesForLanguage(md.parse(data), lang);
+          documentation.renderLinksInText(rootNode);
+
+          // Validate links:
+          {
+            md.visitAll(rootNode, node => {
+              if (!node.text)
+                return;
+              for (const match of node.text.matchAll(/\[(.*?)\]\((.*?)\)/g)) {
+                const [, linkName, linkRef] = match;
+                const isExternal = linkRef.startsWith('http://') || linkRef.startsWith('https://');
+                if (isExternal)
+                  continue
+                if (linkRef.startsWith('./')) {
+                  const linkRefHash = linkRef.indexOf('#');
+                  if (linkRefHash !== -1)
+                    assertFileExists(filePath, linkRef.substring(0, linkRefHash), linkName);
+                  else
+                    assertFileExists(filePath, linkRef, linkName);
+                }
+              }
+            });
+
+            /**
+             * @param {string} filePath 
+             * @param {string} linkRef
+             * @param {string} linkName
+             */
+            function assertFileExists(filePath, linkRef, linkName) {
+              if (path.extname(linkRef) !== '.md')
+                linkRef += '.md';
+              if (linkRef.startsWith('./'))
```

Comment:
let's resolve `../` as well.

Also: let's fix these kind of links - https://github.com/microsoft/playwright/blob/f79fd3a5c24c887e39412bdf74b44a8241421ef5/docs/src/test-reporter-api/class-testcase.md?plain=1#L15

This should be `../test-annotations.md`